docs: connect to Blockchain via RPC (#1545)
tendermint/starport: docs/kb/connect-blockchain.md

---

> description: Connect to Blockchain with RPC

Reviewer: ```suggestion
description: Creating a gRPC Client
```

> description: Connect to Blockchain with RPC

Reviewer: ```suggestion
description: Blockchain Client in Go
```

order: 10
---

> # Connect to Blockchain with RPC

Reviewer: ```suggestion
# Interacting With a Chain Through gRPC
```

> # Connect to Blockchain with RPC

Reviewer: ```suggestion
# Creating a Blockchain Client in Go
```

**Learn how to connect your Blockchain to an independent application with RPC**

> ## Step 1

Reviewer: ```suggestion
## Creating a Blockchain
```

> - **Scaffold a new Blockchain using `starport`**

Reviewer: ```suggestion
Scaffold a new blockchain using `starport`:
```

```zsh

> starport scaffold chain github.com/alijnmerchant21/blog

Reviewer: ```suggestion
starport scaffold chain github.com/cosmonaut/blog
```

```

> - **Scaffold a new list post with title and body**

Reviewer: ```suggestion
Scaffold create, read, update, delete functionality for a type `post` with two fields: `title` and `body`. Use `starport scaffold list` to scaffold code for storing posts in a list-like data structure.
```

```zsh

> starport s list post title body

Reviewer: ```suggestion
starport scaffold list post title body
```

Author: shortcuts are so awesome. Specially `s`

```

> **Awesome! You have scaffolded a Blockchain with list. Lets us start the chain.**

Reviewer: ```suggestion
Start a blockchain node in development:
```

```zsh
starport chain serve
```

> ## Step 2

Reviewer: ```suggestion
## Creating a Blockchain Client
```

> **In a new directory (outside the blockchain directory) create two files - `go.main` and `go.mod`**

Author: Implementing. Sorry for the delay - I was heads down busy with workshop. Implementing all the changes now.

```zsh
touch main.go && touch go.mod
```
**`go.main`**

- **Import files**

```go
package main

import (
	"context"
	"fmt"
	"log"

> 	"github.com/alijnmerchant21/blog/x/blog/types"

Reviewer: ```suggestion
	"github.com/cosmonaut/blog/x/blog/types"
```

	blogtypes "github.com/alijnmerchant21/blog/x/blog/types"
	"github.com/tendermint/starport/starport/pkg/cosmosclient"
)
```

> ***Tip:** Use your own module instead of `github.com/alijnmerchant21/blog/x/blog/types`*

Reviewer: ```suggestion
```

- **Instance of cosmosclient.**

> 

Author: I'm not sure if create an instance is correct. I will have to check this.

```go
func main() {

	cosmos, err := cosmosclient.New(context.Background())
	if err != nil {
		log.Fatal(err)	
	}
```

> - **Using the account name Alice (obtained while scaffolding the blockchain), define the account address**

Reviewer 2: we don't mention Alice account elsewhere, I got a little lost here... am coachable, so teach me!

> - **Using the account name Alice (obtained while scaffolding the blockchain), define the account address**

Author: The Alice we are referring here is the Alice obtained while running Blockchain. It was my mistake to say - `when you scaffold a Blockchain` it should have been - `when you run a blockchain`

```go
accountName := "alice"
address, err := cosmos.Address(accountName)
	if err != nil {
		log.Fatal(err)
	}
```

> - **`Create` a new post, `broadcast` it on the chain and `display` the results**

Author: Yes - It is within `go.main` file.

```go
// Create a post
msg := &types.MsgCreatePost {
	Creator: address.String(),
	Title: "Hello Ali",
	Body: "This is the first post",
}

// Broadcast the post
txResp, err := cosmos.BroadcastTx(accountName, msg)
if err != nil {
	log.Fatal(err)
}
fmt.Print("MsgCreatePost:\n\n")
fmt.Println(txResp)

// Display the results
queryClient := blogtypes.NewQueryClient(cosmos.Context)
queryResp, err := queryClient.Post(context.Background(), &types.QueryGetPostRequest{})

	if err != nil {
		log.Fatal(err)
	}
	fmt.Print("\n\nUser:\n\n")
	fmt.Println(queryResp)

// Tip: To display result of all users, replace POST with POSTALL
queryResp, err := queryClient.PostAll(context.Background(), &types.QueryAllPostRequest{})
```

> - **Add `go.mod` file**

Author: "go. mod file defines the module's module path, which is also the import path used for the root directory, and its dependency requirements, which are the other modules needed for a successful build"

> - **Add `go.mod` file**

Author: Not sure if the title Define client modules is correct.

```go
module github.com/tendermint/starport/local_test/client

go 1.17

require (

> 	github.com/alijnmerchant21/blog v0.0.0-20210825193134-b6859adfa282

Author: when you do `go get github.com/alijnmerchant21/blog` it should give you the version.

Some like this: 

<img width="854" alt="Screen Shot 2021-09-10 at 4 36 31 PM" src="https://user-images.githubusercontent.com/44069404/132914773-08291b9b-30a4-44ce-bf90-bc60da1d7d93.png">

> 	github.com/alijnmerchant21/blog v0.0.0-20210825193134-b6859adfa282

Author: Also, since we are using `cosmonaut` instead of `alijnmerchant21` the version no longer makes sense. Also, this is a `go` thing and not starport thing. So we shouldn't be too concerned about it..

	github.com/tendermint/starport v0.15.0
)

replace github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1

> replace github.com/tendermint/starport => github.com/ilgooz/starport v0.0.500

Reviewer 3: this line should be removed (uses ilgooz/starport) but it's only possible when we make another release in Starport that includes latest changes in cosmosclient pkg.

```

> ***Tip:** Instead of `github.com/alijnmerchant21/blog v0.0.0-20210825193134-b6859adfa282` using your own `blog` package*

Author: `blog package` was referring to the `blog` blockchain created during scaffolding. Agreed, the verbose needs to be changed.

## Step 3
To run this program, you need follow these steps:
- Populate the file `main.go` with the code given above
- Run `go mod tidy`
- Run the main file `go run main.go`

**Output**
```zsh
MsgCreatePost:

Response:
  Height: 3522
  TxHash: E6904A959A4F62B361CAD9C1F8A6976E1B48EB4DA77A1FBB0D41DE921F1E8028
  Data: 0A100A0A437265617465506F737412020801
  Raw Log: [{"events":[{"type":"message","attributes":[{"key":"action","value":"CreatePost"}]}]}]
  Logs: [{"events":[{"type":"message","attributes":[{"key":"action","value":"CreatePost"}]}]}]
  GasWanted: 300000
  GasUsed: 45764

User:

Post:<creator:"cosmos1u27xu76zamjzgus2py87r8kmafea8cp6rvgtv9" 

> title:"Hello Ali" body:"This is the first post" >

Reviewer 2: I'm not sure exactly what we can put here, but let's have a little fun with the first body post?

> title:"Hello Ali" body:"This is the first post" >

Author: This looks good. Though the o/p needs to be changed accordingly. Will do!

```

## Step 4
Similaryly, you can `update` and `delete` the POST
```go

> // Update a Post

Author: adding this to `main.go`
This is updating the post (the post we created in previous steps)

	msgUpd := &types.MsgUpdatePost{
		Creator: address.String(),
		Id:      0,
		Title:   "Hello Ali",

> 		Body:    "This is a modified post",

Author: Sure! We can add some fun!!

	}

> // Delete a Post

Reviewer 2: ````suggestion
```

## Delete a Post

To delete a post:
````

Reviewer 2: again, I would like to tell the user if they are adding more code to the `main.go` file

	msgDel := &types.MsgDeletePost{
		Creator: address.String(),
		Id:      1,
	}
```

> **Final code `main.go`**

Reviewer 2: ```suggestion
## Final Code in main.go 

After completing the steps in this article, your `main.go` file looks like:

```

```go
package main

import (
	"context"
	"fmt"
	"log"

	"github.com/alijnmerchant21/blog/x/blog/types"
	blogtypes "github.com/alijnmerchant21/blog/x/blog/types"
	"github.com/tendermint/starport/starport/pkg/cosmosclient"
)

func main() {

	cosmos, err := cosmosclient.New(context.Background())
	if err != nil {
		log.Fatal(err)	
	}

	accountName := "alice"
	address, err := cosmos.Address(accountName)
	if err != nil {
		log.Fatal(err)
	}

	// Create a post
	msg := &types.MsgCreatePost {
		Creator: address.String(),
		Title: "Hello Ali",
		Body: "This is the first post",
	}

	// Broadcast the post
	txResp, err := cosmos.BroadcastTx(accountName, msg)
	if err != nil {
	log.Fatal(err)
	}
	fmt.Print("MsgCreatePost:\n\n")
	fmt.Println(txResp)

	// Display the results
	queryClient := blogtypes.NewQueryClient(cosmos.Context)
	queryResp, err := queryClient.Post(context.Background(), &types.QueryGetPostRequest{})

	if err != nil {
		log.Fatal(err)
	}
	fmt.Print("\n\nUser:\n\n")
	fmt.Println(queryResp)

> 	// Tip: To display result of all users, replace POST with POSTALL
> 	queryResp, err := queryClient.PostAll(context.Background(), &types.QueryAllPostRequest{})

Reviewer 2: ```suggestion
```

Reviewer 2: instead of a tip, if we want the user to have an option to use POSTALL, let's add this as a step and create a new code entry. 

Let's remove the // Tip: from the code sample, too

Author: Thanks for the suggestion. I'll add `POSTALL` as code

	// Delete a Post
	msgDel := &types.MsgDeletePost{
		Creator: address.String(),
		Id:      1,
	}

	// Broadcast Delete Post
	txResp, err := cosmos.BroadcastTx(accountName, msgDel)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Print("MsgDeletePost:\n\n")
	fmt.Println(txResp)

> 	// Display the results

Reviewer 2: ```suggestion
	// Display the Results
```

Reviewer 2: good job on the comments for each code section, let's be consistent and use title case

	queryClient := blogtypes.NewQueryClient(cosmos.Context)
	queryRespAll, err := queryClient.PostAll(context.Background(), &types.QueryAllPostRequest{})
	if err != nil {
		log.Fatal(err)
	}
	fmt.Print("\n\nUser All:\n\n")
	fmt.Println(queryRespAll)

	// Update a Post
	msgUpd := &types.MsgUpdatePost{
		Creator: address.String(),
		Id:      0,
		Title:   "Hello Ali",
		Body:    "This is a modified post",
	}

	// Broadcast Update Post
	txResp, err := cosmos.BroadcastTx(accountName, msgUpd)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Print("MsgUpdatePost:\n\n")
	fmt.Println(txResp)

> 	// Display the results

Reviewer 2: ```suggestion
	// Display the Results
```
